### PR TITLE
feat: Add cuBLASLt backend for `mm_bf16` and enable multi-tactic autotuning for FP8/MXFP8 runners

### PR DESCRIPTION
**Pull request by vadiklyutiy**
## Summary

- **Add cuBLASLt backend for `mm_bf16`**: new `backend="cublaslt"` option (gated to SM100/SM103). Autotuning across all available cuBLASLt algorithms via `get_valid_tactics()`. 
- **Enable multi-tactic autotuning for single-tactic GEMM runners**: `CublasFp8GemmRunner`, `CudnnFp8GemmRunner`, and `CudnnMxfp8GemmRunner` previously hardcoded a single tactic (`return [0]` / `return [-1]`), preventing the autotuner from exploring better algorithms (the same are done for FP4 and FP16). Now all three enumerate available algorithms/plans via `get_valid_tactics()` and pass the selected tactic through to execution.
- **Improve test coverage**: added `cublaslt` and `auto` backends + `auto_tuning` parameter to `test_mm_bf16.py`, `auto` backend to `test_bmm_bf16.py`, and a dedicated edge-case test for zero-algorithm handling.
## Test Results

All GEMM/BMM test suites pass with 0 failures:

| Test File | Passed | Failed |
|-----------|--------|--------|
| `test_mm_bf16.py` | 1441 | 0 |
| `test_mm_fp8.py` | 30 | 0 |
| `test_mm_fp4.py` | 1440 | 0 |
| `test_mm_mxfp8.py` | 1843 | 0 |
| `test_bmm_bf16.py` | 144 | 0 |
| `test_bmm_fp8.py` | 1188 | 0 |
| `test_bmm_mxfp8.py` | 288 | 0 |
| **Total** | **6374** | **0** |

Test additions:
- `test_mm_bf16.py`: added `cublaslt` and `auto` to backend parametrize, added `auto_tuning` parameter, added `test_cublaslt_bf16_runner_zero_algos` edge-case test.
- `test_bmm_bf16.py`: added `auto` backend.

## Summary by CodeRabbit

* **New Features**
  * Added a cuBLASLt BF16 GEMM backend with selectable algorithm tactics; module generated for supported hardware.
  * Exposed APIs to enumerate and run FP8/BF16 GEMM algorithms for explicit tactic selection.
  * Expanded backend choices with "cublaslt" and broader "auto" routing.

* **Bug Fixes**
  * Robustified autotuner hashing to avoid failures with unhashable attributes.

* **Tests**
  * Extended tests for new backends, autotuning options, auto routing, and zero-algorithm failure handling.

### CODE REVIEW

File: flashinfer/gemm/gemm_base.py [hunk 1]
```diff
@@ -870,6 +960,108 @@ def forward(
...
+                # cuBLASLt with BF16 inputs supports BF16 or FP32 output natively.
+                # FP16 output is achieved via BF16 compute + cast.
+                if out_dtype == torch.float16:
+                    return torch.bfloat16
+                return out_dtype
+
+            def _get_algos(self, inputs):
+                a, b, _, _, out, workspace_buffer = inputs
+                compute_dt = self._compute_dtype(out.dtype)
+                key = (a.shape[0], b.shape[0], a.shape[1], compute_dt)
```
[the remark this answers is not in the document]
Comment (vadiklyutiy) [reply flashinfer/gemm/gemm_base.py]: fixed

File: tests/gemm/test_mm_bf16.py [hunk 2]
```diff
@@ -68,12 +79,41 @@ def test_mm_bf16(
...
+    compute_capability = get_compute_capability(torch.device("cuda"))
+    cc_num = compute_capability[0] * 10 + compute_capability[1]
+    if not mm_bf16.is_backend_supported("cublaslt", cc_num):
+        pytest.skip("cublaslt backend not supported on this GPU")
+
+    runner = get_mm_bf16_cublaslt_module().cublaslt_bf16_gemm_runner()
+
+    m, n, k = 16, 1024, 1024
+    a = torch.randn(m, k, device="cuda", dtype=torch.bfloat16)
+    b = torch.randn(n, k, device="cuda", dtype=torch.bfloat16)
```
[the remark this answers is not in the document]
Comment (vadiklyutiy) [reply tests/gemm/test_mm_bf16.py]: fixed

File: csrc/mm_bf16_cublaslt.cu [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  cudaDataType_t d_type = get_d_type(out.dtype());
+
+  ffi::CUDADeviceGuard device_guard(mat1.device().device_id);
+  auto lt_handle = reinterpret_cast<cublasLtHandle_t>(cublas_handle);
+  int max_algos = static_cast<int>(algo_buffer.numel() * get_element_size(algo_buffer) /
+                                   flashinfer::mm_bf16_cublaslt::kAlgoBytes);
+  return static_cast<int64_t>(flashinfer::mm_bf16_cublaslt::get_algorithms(
+      static_cast<int>(m), static_cast<int>(n), static_cast<int>(k), d_type,
+      workspace_buffer.numel() * get_element_size(workspace_buffer), lt_handle,
+      algo_buffer.data_ptr(), max_algos));
```
[the remark this answers is not in the document]
Comment (vadiklyutiy) [reply csrc/mm_bf16_cublaslt.cu:78]: fixed

File: flashinfer/autotuner.py [hunk 4]
```diff
@@ -397,7 +397,14 @@ def forward(
...
     def __hash__(self):
-        return hash(tuple(self.__dict__.values()))
+        hashable_vals = []
+        for v in self.__dict__.values():
+            try:
+                hash(v)
+                hashable_vals.append(v)
+            except TypeError:
+                hashable_vals.append(id(v))
+        return hash(tuple(hashable_vals))
```
[the remark this answers is not in the document]
Comment (vadiklyutiy) [reply flashinfer/autotuner.py:422]: fixed

File: flashinfer/gemm/gemm_base.py [hunk 5]
```diff
@@ -353,21 +436,22 @@ def mm_bf16(
...
 
     pdl: bool
-        Whether to use persistant data loader mode. Enabled for TGV backend. Defaults to ``False``.
+        Whether to enable Programmatic Dependent Launch (PDL). Enabled for TGV backend. Defaults to ``False``.
 
     out: Optional[torch.Tensor]
-        Out tensor, shape (m, n), bf16, fp16, or fp32. Enabled for CUTLASS and cuDNN backends.
-        Defaults to ``None``.
+        Out tensor, shape (m, n), bf16, fp16, or fp32. Enabled for CUTLASS, cuDNN, and cuBLASLt
+        backends. Defaults to ``None``.
```
[the remark this answers is not in the document]
Comment (vadiklyutiy) [reply flashinfer/gemm/gemm_base.py]: fixed

**Comment by vadiklyutiy** [comment]
> [FAILED] Pipeline [#47392419](https://gitlab-master.nvidia.com/dl/flashinfer/flashinfer-ci/-/pipelines/47392419): 6/20 passed

I looked in several fails. It looks that fails don't relate to this PR - fails in norm, quantisation, topk, etc

@dhiraj113 How do you usually process in such cases?

**Comment by dhiraj113** [comment]
@vadiklyutiy Could you resolve the conflicts and rebase/merge with `main` branch again. There is a lot of diff that is showing that I don't think is from your PR.

**Comment by vadiklyutiy** [comment]
> @vadiklyutiy Could you resolve the conflicts and rebase/merge with `main` branch again. There is a lot of diff that is showing that I don't think is from your PR.

done

File: benchmarks/routines/gemm.py [hunk 6]
```diff
@@ -1618,7 +1627,7 @@ def testMmBf16(args):
         return res
 
     def run_backend(backend, a, b, bias, use_pdl, out_dtype):
-        if backend in ["cudnn", "cutlass", "tgv", "auto"]:
+        if backend in ["cudnn", "cutlass", "tgv", "cublaslt", "auto"]:
```
Comment (dhiraj113) [inline benchmarks/routines/gemm.py:1654]: Should adding this backend be done for `testBmmBf16` too?

File: include/flashinfer/gemm/bmm_fp8.cuh [hunk 7]
```diff
@@ -19,6 +19,8 @@
 #include <cublasLt.h>
 #include <cuda_fp8.h>
 
+#include <array>
+#include <cstring>
```
Comment (dhiraj113) [inline include/flashinfer/gemm/bmm_fp8.cuh:23]: For the sake of symmetry, should this file have been named with a `_cublas` suffix, just like `mm_bf16_cublaslt.cuh`.

File: include/flashinfer/gemm/mm_bf16_cublaslt.cuh [hunk 8]
```diff
@@ -0,0 +1,130 @@
+/*
```
Comment (dhiraj113) [inline include/flashinfer/gemm/mm_bf16_cublaslt.cuh:1]: Why the support has been added only for `mm`, and not for both `mm` and `bmm`? Is it easier to add support only for `b = 1` case, or was it the only requirement for your usecase?

File: flashinfer/autotuner.py [hunk 9]
```diff
@@ -397,7 +397,16 @@ def forward(
         raise NotImplementedError
 
     def __hash__(self):
-        return hash(tuple(self.__dict__.values()))
+        hashable_vals = []
```
Comment (dhiraj113) [inline flashinfer/autotuner.py:413]: Why is this new hash function implemented? What's the rationale?

File: flashinfer/gemm/gemm_base.py [hunk 10]
```diff
@@ -126,11 +165,33 @@ def forward(
...
+                    cublas_handle = torch.cuda.current_blas_handle()
+                if tactic >= 0:
+                    algo_buf, count = self._get_algos(inputs)
+                    if count == 0:
+                        raise RuntimeError(
+                            "cuBLASLt heuristic returned zero FP8 algorithms for "
+                            f"A={tuple(a.shape)}, B={tuple(b.shape)}, out={tuple(out.shape)}."
+                        )
+                    if tactic >= count:
+                        tactic = 0
```
Comment (dhiraj113) [inline flashinfer/gemm/gemm_base.py]: Wouldn't it be better to return an error here?

File: flashinfer/gemm/gemm_base.py [hunk 11]
```diff
@@ -6999,6 +7273,7 @@ def _get_cudnn_mxfp8_gemm_graph(
     out: Optional[torch.Tensor] = None,
     block_size: int = 32,  # mxfp8 block size is 32
     tactic: int = -1,
+    build_all: bool = False,
```
Comment (dhiraj113) [inline flashinfer/gemm/gemm_base.py]: Can you replace `build_all` with `policy` as done for `build_cudnn_gemm_fp4_graph`.

File: benchmarks/routines/gemm.py (hunk 6 above)
Comment (vadiklyutiy) [reply benchmarks/routines/gemm.py:1654]: This changes intentionally don't touch bmm_bf16

File: include/flashinfer/gemm/bmm_fp8.cuh (hunk 7 above)
Comment (vadiklyutiy) [reply include/flashinfer/gemm/bmm_fp8.cuh:23]: yes, but prefer to avoid in in this review to make the diff smaller (diff is a little bit big now)

File: include/flashinfer/gemm/mm_bf16_cublaslt.cuh (hunk 8 above)
Comment (vadiklyutiy) [reply include/flashinfer/gemm/mm_bf16_cublaslt.cuh:1]: 2 reason
 - limit the scope of changes
 - for my case we need only `mm` variant

File: flashinfer/autotuner.py (hunk 9 above)
Comment (vadiklyutiy) [reply flashinfer/autotuner.py:413]: The old __hash__  breaks now because CublasFp8GemmRunner and CublasltBf16GemmRunner (both TunableRunner subclasses) have a new self._algo_cache: dict = {} attribute - and dict is not hashable. The new implementation skips *_cache fields and falls back to id(v) for remaining unhashable values.

Added a concise comment explaining the rationale directly in the code.

File: flashinfer/gemm/gemm_base.py (hunk 10 above)
Comment (vadiklyutiy) [reply flashinfer/gemm/gemm_base.py]: added error

File: flashinfer/gemm/gemm_base.py (hunk 11 above)
Comment (vadiklyutiy) [reply flashinfer/gemm/gemm_base.py]: done

File: flashinfer/gemm/gemm_base.py [hunk 12]
```diff
@@ -108,16 +114,49 @@ def _match_sm_version(device: torch.device, sm_version: list[str]):
...
+            def __init__(self):
+                self._algo_cache: dict = {}
+
+            def get_cache_key_extras(self, inputs: List[torch.Tensor]) -> tuple:
+                a, b, _, _, out, _ = inputs
+                return (a.dtype, b.dtype, out.dtype)
+
+            def _get_algos(self, inputs):
+                a, b, scale_a, scale_b, out, workspace_buffer = inputs
+                key = (a.shape, b.shape, a.dtype, b.dtype, out.dtype)
```
Comment (dhiraj113) [inline flashinfer/gemm/gemm_base.py]: Can you help me understand why this `key` differs from the above return value in `get_cache_key_extras`?

Comment (vadiklyutiy) [reply flashinfer/gemm/gemm_base.py]: no big difference 
I unified this place

File: flashinfer/gemm/gemm_base.py [hunk 13]
```diff
@@ -870,6 +975,125 @@ def forward(
...
+                tactic: int = -1,
+                do_preparation: bool = False,
+                **kwargs,
+            ) -> torch.Tensor:
+                a, b, _, _, out, workspace_buffer = inputs
+                with torch.cuda.device(a.device):
+                    cublas_handle = torch.cuda.current_blas_handle()
+                b_t = b.transpose(-2, -1)
+
+                need_cast = out.dtype == torch.float16
```
Comment (dhiraj113) [inline flashinfer/gemm/gemm_base.py]: Doesn't converting `bf16` to half cause greater precision loss than converting `fp32` to `half` . So, if we have to be faithful here, the output of cublasLt should be `fp32` and not `bf16`.

Comment (vadiklyutiy) [reply flashinfer/gemm/gemm_base.py]: good catch! fixed

**Comment by dhiraj113** [comment]
@gemini-code-assist - Can you summarize the code review and changes done so far

**Review by dhiraj113 — APPROVED** [review]
Thanks for the contribution.

**Comment by vadiklyutiy** [comment]
@dhiraj113 could you pls advise how we can process with merging?

**Comment by vadiklyutiy** [comment]
@dhiraj113 could you help me to understand why auto-merge haven't worked?

**Review by jimmyzho — APPROVED** [review]
LGTM. Will merge once internal/public CI is green

**Comment by vadiklyutiy** [comment]
@flashinfer-bot run

**Comment by vadiklyutiy** [comment]
@jimmyzho could you pls run CI

**Comment by vadiklyutiy** [comment]
> /bot run

@dhiraj113  Could you retry failed? I looked into logs and seems all fails are infra problems

**Comment by vadiklyutiy** [comment]
@dhiraj113
1. PR Label Cleanup not related
2. JIT Unittest 1(A10G) - reran OK
3. JIT Unittest (T4) - reran OK
4. AOT Build Import (x64, cu130) - Docker runner died mid-build at ninja [2139/6448]: time="..." level=error msg="error waiting for container: unexpected EOF" → exit code 125. Build was healthy until then (only normal ptxas advisories). - infra fail

**Comment by vadiklyutiy** [comment]
Can we count it as "green CI"?
